### PR TITLE
Track hash changes in session history

### LINKED ISSUE #13437: Navigating to the same URL with a fragment needs to interact with session history
Currently, `ScriptThread::handle_navigate` short-circuits if the pre-fragment URL is the same. It scrolls to the chosen fragment then returns without performing any actual load. This means that the session history in the constellation doesn't know anything about the navigation.

cc @ConnorGBrewster for suggestions

### LINKED ISSUE #14970: hashchange DOM event should be dispatched when the document location's hash is changed
1) Open https://lhorie.github.io/todomvc-perf-comparison/todomvc-benchmark/ in Servo.
2) Select "Run all"

The tests will start running. At some point during the benchmark, execution stops and nothing further happens. The rendering is still running (can be confirmed with -Z wr-stats).

*Sometimes* you get several JS errors written on the console - but this doesn't always happen.

For example:

```
Om,Adding50Items,expected,50,got,0
Om,CompletingAllItems,expected,50,got,0
ERROR:script::dom::bindings::error: Error at https://lhorie.github.io/todomvc-perf-comparison/todomvc-benchmark/todomvc/ractive/bower_components/ractive/Ractive.js:3:19921 Could not parse template: ...
							&lt;!-- The <- failed at character 1108 ->{{&gt;item}} partial...
ERROR:script::dom::bindings::error: Error at https://lhorie.github.io/todomvc-perf-comparison/todomvc-benchmark/todomvc/ractive/js/persistence.js:45:1 todoList is not defined
ERROR:script::dom::bindings::error: Error at https://lhorie.github.io/todomvc-perf-comparison/todomvc-benchmark/todomvc/ractive/js/routes.js:25:1 todoList is not defined

```

### PR DESCRIPTION
**Pull request by cbrewster**
Adds tracking of hash changes in the session history.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14970 fix #13437 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21048)

### CODE REVIEW

File: tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/hashchange_event.html.ini [hunk 1]
```diff
@@ -1,6 +1,4 @@
 [hashchange_event.html]
-  type: testharness
-  expected: TIMEOUT
   [Queue a task to fire hashchange event]
-    expected: TIMEOUT
+    expected: FAIL
```
Comment (jdm) [inline tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/hashchange_event.html.ini]: Why do we fail this test?

Comment (jdm) [reply tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/hashchange_event.html.ini]: Interestingly, Firefox passes both tests.

Comment (jdm) [reply tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/hashchange_event.html.ini]: Also it looks like bubbles should be false per the "Example" block at https://dom.spec.whatwg.org/#concept-event-fire.

Comment (jdm) [reply tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/hashchange_event.html.ini]: Given https://github.com/web-platform-tests/wpt/pull/11355 from 9 days ago, I think we should update 004.html to match.

**Comment by paulrouget** [comment]
Does this trigger the `HistoryChanged` embedder message?

**Comment by cbrewster** [comment]
@paulrouget looks like it doesn't. I don't think `history.pushState` does either, so I will need to fix both of those.

File: tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/hashchange_event.html.ini (hunk 1 above)
Comment (cbrewster) [reply tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/hashchange_event.html.ini]: Ahh that explains it! I'll update it to `false`

**Comment by asajeffrey** [comment]
OK, this looks really good! Squash, rebase and r=me.

**Comment by jdm** [comment]
```
error[E0061]: this function takes 1 parameter but 0 parameters were supplied
    --> components/script/dom/window.rs:1612:39
     |
1087 |     pub fn task_canceller(&self, name: TaskSourceName) -> TaskCanceller {
     |     ------------------------------------------------------------------- defined here
...
1612 |                         Box::new(self.task_canceller().wrap_task(task)),
     |                                       ^^^^^^^^^^^^^^ expected 1 parameter

error: aborting due to previous error
```

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /css/cssom-view/scroll-behavior-smooth.html:
  └ PASS [expected NOTRUN] Smooth scrolling while doing history navigation.

  ▶ Unexpected subtest result in /css/cssom-view/scroll-behavior-smooth.html:
  │ FAIL [expected TIMEOUT] Instant scrolling while doing history navigation.
  │   → assert_equals: Shouldn't be scrolled back to top yet. expected 0 but got 28739
  │ 
  │ instant/window.onhashchange/window.onhashchange/<@http://web-platform.test:8000/css/cssom-view/scroll-behavior-smooth.html:80:11
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  └ instant/window.onhashchange/window.onhashchange@http://web-platform.test:8000/css/cssom-view/scroll-behavior-smooth.html:78:9
```

**Comment by cbrewster** [comment]
@bors-servo retry
 * websockets

**Pull request merged**